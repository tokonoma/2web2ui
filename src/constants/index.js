/* eslint-disable max-lines */
import { tokens } from '@sparkpost/design-tokens-hibana';

export const DEFAULT_REDIRECT_ROUTE = '/landing';
export const AFTER_JOIN_REDIRECT_ROUTE = '/onboarding/plan';
export const SIGN_UP_ROUTE = '/join';
export const AUTH_ROUTE = '/auth';
export const TFA_ROUTE = '/auth/tfa';
export const ENABLE_TFA_AUTH_ROUTE = '/auth/enable-tfa';
export const SSO_AUTH_ROUTE = '/auth/sso';
export const DASHBOARD_ROUTE = '/dashboard';
export const ONBOARDING_SENDINGDOMAIN_ROUTE = '/onboarding/sending-domain';
export const COOKIE_DOMAIN = '.sparkpost.com';

export const FORMATS = {
  DATE: 'MMM Do',
  LONG_DATE: 'MMM Do YYYY',
  SHORT_DATE: 'YYYY-MM-DD',

  TIME: 'h:mma',
  MILITARY_TIME: 'H:mm',
  ANTE_MILITARY_TIME: 'H:mma', // for 24 hour accidentally including am/pm (FE-61)

  DATETIME: 'MMM Do h:mma',
  LONG_DATETIME: 'MMM Do YYYY h:mma',
  LONG_DATETIME_ALT: 'MMM DD, YYYY [at] h:mma',

  INPUT_DATES: ['YYYY-MM-DD'],
  INPUT_TIMES: ['h:mma', 'H:mm', 'H:mma'],
  INPUT_DATETIMES: ['YYYY-MM-DD h:mma', 'YYYY-MM-DD H:mm', 'YYYY-MM-DD H:mma'],
};

export const RELATIVE_DATE_OPTIONS = ['hour', 'day', '7days', '10days', 'custom'];

export const LINKS = {
  SPC_EU_URL: 'https://app.eu.sparkpost.com',
  SPC_US_URL: 'https://app.sparkpost.com',
  SP_HOME_PAGE: 'https://www.sparkpost.com',
  API_DOCS: 'https://developers.sparkpost.com/api',
  DOCS: 'https://www.sparkpost.com/docs',
  PREMIUM_SUPPORT: 'https://www.sparkpost.com/contact-premium',
  ENTERPRISE_SUPPORT: 'https://www.sparkpost.com/contact-enterprise',
  RECIP_API: 'https://developers.sparkpost.com/api/recipient-lists.html',
  AB_TESTING_API: 'https://developers.sparkpost.com/api/ab-testing.html',
  SUBACCOUNTS: 'https://sparkpost.com/docs/user-guide/subaccounts/',
  SUBACCOUNTS_API: 'https://developers.sparkpost.com/api/subaccounts.html',
  TOU: 'https://www.sparkpost.com/policies/tou',
  DOMAIN_VERIFICATION:
    'https://www.sparkpost.com/docs/tech-resources/enabling-multiple-custom-tracking-domains',
  SENDING_REQS: 'https://www.sparkpost.com/docs/getting-started/requirements-for-sending-domains',
  SENDING_SETUP: 'https://www.sparkpost.com/docs/getting-started/setting-up-domains/',
  TFA_BACKUP_CODES: 'https://support.sparkpost.com/customer/portal/articles/1948449',
  IP_WARM_UP: 'https://support.sparkpost.com/customer/portal/articles/1972209-ip-warm-up-overview',
  DAILY_USAGE: 'https://www.sparkpost.com/docs/getting-started/what-counts-daily-monthly-usage',
  LEARN_MORE_TFA:
    'https://www.sparkpost.com/docs/my-account-and-profile/enabling-two-factor-authentication',
  ONBOARDING_SENDING:
    'https://www.sparkpost.com/docs/getting-started/getting-started-sparkpost/#preparing-your-from-address',
  RECAPTCHA_LIB_URL: 'https://www.google.com/recaptcha/api.js',
  DAILY_MONTHLY_QUOTA_LIMIT_DOC: 'https://support.sparkpost.com/customer/portal/articles/2030894',
  GETTING_STARTED_GUIDE: 'https://www.sparkpost.com/docs/getting-started/getting-started-sparkpost',
  ADMIN_BOUNCE: 'https://support.sparkpost.com/', // TODO Add this to the bounce page
  SSO_GUIDE: 'https://www.sparkpost.com/docs/my-account-and-profile/sso',
  ALERTS_DOCS: 'https://developers.sparkpost.com/api/alerts/#alerts-post-create-an-alert',
  ALERTS_SURVEY: 'https://goo.gl/forms/rnIuTvdVF2xhpKCy2',
  SNIPPET_API_DOCS: 'https://developers.sparkpost.com/api/snippets/',
  MANDATORY_TFA:
    'https://www.sparkpost.com/docs/my-account-and-profile/enabling-two-factor-authentication/',
  RECIPIENT_VALIDATION_ACCESS: 'https://www.sparkpost.com/rv-access',
  AUTO_IP_WARMUP_SETUP: 'https://www.sparkpost.com/docs/user-guide/automated-ip-warmup/',
  MAILBOX_PROVIDERS: 'https://www.sparkpost.com/docs/user-guide/alerts/#mailbox-providers',
};

export const ENTERPRISE_PLAN_CODES = ['ent1'];

export const AWS_COOKIE_NAME = 'aws-mkt';

export const DAILY_LIMIT_REQUEST_TEMPLATE = 'daily-limit-increase';

export const SPC_EU_TENANT = 'spceu';
export const SPC_TENANT = 'spc';

export const CROSS_LINK_MAP = {
  [SPC_TENANT]: { label: '', url: LINKS.SPC_US_URL },
  [SPC_EU_TENANT]: { label: 'EU', url: LINKS.SPC_EU_URL },
};

export const TYPEAHEAD_LIMIT = 100;
export const METRICS_API_LIMIT = 1000;

export const FORMS = {
  LOGIN: 'loginForm',
  SSO: 'ssoLoginForm',
  JOIN: 'joinForm',
  JOIN_PLAN: 'joinPlanForm',
  JOIN_SENDING_DOMAIN: 'joinSendingDomainForm',
  INVITE_USER: 'userInviteForm',
  EDIT_USER: 'userEditForm',
  EVENTS_SEARCH: 'eventsSearchForm',
  RV_ADDPAYMENTFORM: 'rvAddPaymentForm',
  RV_ADDPAYMENTFORM_UPLOADLISTPAGE: 'rvAddPaymentFormUploadListPage',
};

//These filters are the searchable filters in the 'more filters' modal on the Events Search page
export const EVENTS_SEARCH_FILTERS = {
  recipient_domains: {
    placeholder: 'list of domain parts (e.g., gmail, yahoo.co, yahoo.co.jp, .fr)',
    label: 'Recipient Domains',
  },
  from_addresses: {
    placeholder: 'list of full from addresses (e.g., john@mycompany.com)',
    label: 'From Addresses',
  },
  sending_domains: {
    placeholder: 'list of domain parts (e.g., joespizza, marketing.joespizza)',
    label: 'Sending Domains',
  },
  subjects: {
    placeholder: 'list of full words (e.g. happy birthday, christmas)',
    label: 'Subjects',
  },
  bounce_classes: {
    placeholder: 'list of bounce classification codes (e.g. 10, 30)',
    label: 'Bounce Classes',
  },
  reasons: {
    placeholder: 'list of full words in raw reasons (e.g. 5.4.7, IP, Mail rejected)',
    label: 'Reasons',
  },
  campaigns: {
    placeholder: 'list of full words in campaign ids (e.g. newsletter, fire-sale)',
    label: 'Campaign IDs',
  },
  templates: {
    placeholder: 'list of full words in template ids (e.g. newsletter, fire-sale)',
    label: 'Template IDs',
  },
  sending_ips: { placeholder: 'list of ip addresses (e.g., 10.0.0.4.25)', label: 'Sending IPs' },
  ip_pools: {
    placeholder: 'list of full words in ip pool id (e.g billing, administrative)',
    label: 'IP Pools',
  },
  subaccounts: { placeholder: 'list of subaccount ids (e.g. 346, 823)', label: 'Subaccount IDs' },
  messages: {
    placeholder: 'list of message ids (e.g. 00129004225c33cc4c45)',
    label: 'Message IDs',
  },
  transmissions: {
    placeholder: 'list of transmission ids (e.g. 337221619334066426)',
    label: 'Transmissions IDs',
  },
  ab_tests: {
    placeholder: 'list of full words in ab test ids (e.g. newsletter, fire-sale)',
    label: 'AB Test IDs',
  },
  ab_test_versions: {
    placeholder: 'list of version numbers; requires AB Test ID filter(e.g. 4)',
    label: 'AB Test Versions',
  },
};

export const ALL_EVENTS_FILTERS = {
  recipients: { placeholder: 'Filter by recipient email address', label: 'Recipients' },
  events: { label: 'Event' },
  ...EVENTS_SEARCH_FILTERS,
};

export const DEFAULT_PER_PAGE_BUTTONS = [10, 25, 50, 100];

export const ANALYTICS_CREATE_ACCOUNT = 'create account';
export const ANALYTICS_ADDON_IP = 'dedicated_ips';
export const ANALYTICS_PREMIUM_SUPPORT = 'premium-support';
export const ANALYTICS_ENTERPRISE_SUPPORT = 'enterprise-support';
export const ANALYTICS_ONBOARDING = 'onboarding';
export const ANALYTICS_ONBOARDING_CREATE_DOMAIN = 'create-sending-domain-verify-';
export const ANALYTICS_WHITELISTED_FORMS = [
  FORMS.LOGIN,
  FORMS.JOIN,
  FORMS.JOIN_PLAN,
  FORMS.JOIN_SENDING_DOMAIN,
];

export const MAILBOX_PROVIDERS = {
  apple: 'Apple',
  gmail: 'Gmail',
  google_apps: 'Google Apps',
  microsoft: 'Microsoft',
  verizon_media: 'Verizon Media',
  yahoo_uk: 'Yahoo UK',
  us_broadband: 'US Broadband',
  uk_providers: 'UK Providers',
  italian_providers: 'Italian Providers',
  russian_providers: 'Russian Providers',
  french_providers: 'French Providers',
  chinese_providers: 'Chinese Providers',
  german_providers: 'German Providers',
  canadian_providers: 'Canadian Providers',
  japanese_providers: 'Japanese Providers',
  other_european_providers: 'Other European Providers',
  other: 'Other Providers',
};

export const PLAN_TIERS = {
  test: 'Test Account',
  starter: 'Starter',
  premier: 'Premier',
};

export const RECIPIENT_VALIDATION_TIERS = [
  { volumeMin: 0, volumeMax: 5000, cost: 0.01, displayedCost: '$0.010' },
  { volumeMin: 5000, volumeMax: 10000, cost: 0.008, displayedCost: '$0.008' },
  { volumeMin: 10000, volumeMax: 50000, cost: 0.006, displayedCost: '$0.006' },
  { volumeMin: 50000, volumeMax: 100000, cost: 0.004, displayedCost: '$0.004' },
  { volumeMin: 100000, volumeMax: 250000, cost: 0.003, displayedCost: '$0.003' },
  { volumeMin: 250000, volumeMax: 750000, cost: 0.0015, displayedCost: '$0.0015' },
  { volumeMin: 750000, volumeMax: 1000000, cost: 0.001, displayedCost: '$0.0010' },
  { volumeMin: 1000000, volumeMax: Infinity, cost: 0.00075, displayedCost: '$0.00075' },
];

export const HIBANA_METRICS_COLORS = [
  tokens.color_red_700,
  tokens.color_purple_800,
  tokens.color_yellow_500,
  tokens.color_blue_700,
  tokens.color_teal_600,
  tokens.color_magenta_800,
  tokens.color_blue_900,
  tokens.color_green_800,
  tokens.color_red_600,
  tokens.color_purple_600,
  tokens.color_yellow_700,
  tokens.color_teal_700,
  tokens.color_magenta_700,
  tokens.color_green_900,
];

export { ROLES, SUBACCOUNT_ROLES, ROLE_LABELS } from './users';
export { CAMPAIGN_IDS } from './vwo';
