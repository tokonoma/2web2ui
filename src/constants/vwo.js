export const CAMPAIGN_IDS = {
  ONBOARDING_SENDINGDOMAIN: [2, 19],
  CREATE_SENDINGDOMAIN: [3, 19],
  VERIFY_SENDINGDOMAIN: [203],
};
