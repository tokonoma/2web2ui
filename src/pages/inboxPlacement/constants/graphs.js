import { tokens } from '@sparkpost/design-tokens-hibana';

// TODO: Once OG theme is removed, remove theme-specific objects containing color configuration
export const OG_PERCENT_WHEEL = {
  color: '#0CBAC7',
  background: '#EEEEEE',
};

export const HIBANA_PERCENT_WHEEL = {
  color: tokens.color_blue_700,
  background: tokens.color_gray_300,
};
