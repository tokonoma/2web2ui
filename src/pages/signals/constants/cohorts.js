import { tokens } from '@sparkpost/design-tokens-hibana';

export default {
  new: {
    OGFill: '#88E2E7',
    hibanaFill: tokens.color_blue_500,
    OGStroke: '#88E2E7',
    hibanaStroke: tokens.color_blue_500,
    label: 'New',
    description: 'Never engaged, first email in last 7 days',
  },
  uneng: {
    OGFill: '#50CFDA',
    hibanaFill: tokens.color_blue_600,
    OGStroke: '#50CFDA',
    hibanaStroke: tokens.color_blue_600,
    label: 'Never Engaged',
    description: '365+ days since last engagement',
  },
  '365d': {
    OGFill: '#29B9C7',
    hibanaFill: tokens.color_blue_700,
    OGStroke: '#29B9C7',
    hibanaStroke: tokens.color_blue_700,
    label: 'Not Recently Engaged',
    description: '90-365 days since last engagement',
  },
  '90d': {
    OGFill: '#219EA8',
    hibanaFill: tokens.color_blue_800,
    OGStroke: '#219EA8',
    hibanaStroke: tokens.color_blue_800,
    label: 'Semi Recently Engaged',
    description: '14-90 days since last engagement',
  },
  '14d': {
    OGFill: '#1A838B',
    hibanaFill: tokens.color_blue_900,
    OGStroke: '#1A838B',
    hibanaStroke: tokens.color_blue_900,
    label: 'Recently Engaged',
    description: '0-14 days since last engagement',
  },
};
