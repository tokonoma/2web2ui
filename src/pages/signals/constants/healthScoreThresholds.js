import React from 'react';
import { tokens } from '@sparkpost/design-tokens-hibana';
import { CheckCircleOutline, ErrorOutline } from '@sparkpost/matchbox-icons';

const thresholds = {
  danger: {
    color: tokens.color_red_700,
    barColor: tokens.color_red_700,
    icon: ErrorOutline,
    condition: n => n < 55,
    description: (
      <span data-id="health-score-below-55-description">
        A Health score below 55 is considered to be <strong>bad</strong>.
      </span>
    ),
  },
  warning: {
    color: tokens.color_yellow_400,
    barColor: tokens.color_yellow_400,
    icon: ErrorOutline,
    condition: n => n < 80 && n >= 55,
    description: (
      <span data-id="health-score-below-80-description">
        A Health score below 80 <strong>needs improvement</strong>.
      </span>
    ),
  },
  good: {
    color: tokens.color_green_800,
    barColor: tokens.color_green_800,
    icon: CheckCircleOutline,
    condition: n => n >= 80,
    description: (
      <span data-id="health-score-above-80-description">
        A Health score above 80 is considered to be <strong>good</strong>.
      </span>
    ),
  },
};

export default thresholds;

export const HEALTH_SCORE_COLORS_V3 = {
  danger: '#D1E4F4',
  warning: '#91C5FD',
  good: '#4194ED',
};
