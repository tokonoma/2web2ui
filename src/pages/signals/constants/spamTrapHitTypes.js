import { rekey, toCollection } from 'src/helpers/object';
import { tokens } from '@sparkpost/design-tokens-hibana';

// TODO: Remove multiple fill values when OG theme is removed
const spamTrapHitTypes = {
  trap_hits_parked: {
    OGFill: '#50CFDA',
    hibanaFill: tokens.color_blue_900,
    label: 'Parked',
    description: `
      This is an email address from a domain hosted by a parked domain provider. These addresses
      can point to list quality issues but do not affect your reputation.
    `,
  },
  trap_hits_recycled: {
    OGFill: '#29B9C7',
    hibanaFill: tokens.color_blue_700,
    label: 'Recycled',
    description: `
      An email address that once was a real address, but now is a spam trap. These addresses signal
      that you should clean out old unengaged recipients.
    `,
  },
  trap_hits_typo: {
    OGFill: '#1A838B',
    hibanaFill: tokens.color_blue_500,
    label: 'Typo',
    description: `
      This is an email address from a domain that looks like a real mailbox provider, like
      gmal.com. These addresses signal that you should improve your list acquisition practices.
    `,
  },
};

export const spamTrapHitTypesCollection = toCollection(spamTrapHitTypes);
export const spamTrapHitTypesByLabel = rekey(spamTrapHitTypes, 'label');

export default spamTrapHitTypes;
