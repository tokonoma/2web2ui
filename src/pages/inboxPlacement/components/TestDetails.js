import React, { useState, useCallback } from 'react';
import { format } from 'date-fns';
import _ from 'lodash';
import { Search } from '@sparkpost/matchbox-icons';
import { FORMATS } from 'src/constants';
import { OGOnlyWrapper } from 'src/components/hibana';
import { Grid, Panel, Select, Tag, TextField } from 'src/components/matchbox';
import { Bold } from 'src/components/text';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import FolderPlacementBarChart from './FolderPlacementBarChart';
import PlacementBreakdown from './PlacementBreakdown';
import TimeToReceiveSection from './TimeToReceiveSection';
import AuthenticationResults from './AuthenticationResults';
import { PLACEMENT_FILTER_TYPES, PLACEMENT_FILTER_LABELS } from '../constants/types';
import OGStyles from './TestDetails.module.scss';
import HibanaStyles from './TestDetailsHibana.module.scss';

const PLACEMENTS_TYPE_OPTIONS = Object.values(PLACEMENT_FILTER_TYPES).map(type => ({
  label: PLACEMENT_FILTER_LABELS[type],
  value: type,
}));

const TestDetails = ({
  details,
  placementsByProvider,
  placementsByRegion,
  placementsBySendingIp,
}) => {
  const styles = useHibanaOverride(OGStyles, HibanaStyles);
  const [breakdownType, setBreakdownType] = useState(PLACEMENTS_TYPE_OPTIONS[0].value);
  const [searchPlacements, setSearchPlacements] = useState('');

  const debouncedSetSearchPlacements = useCallback(
    _.debounce(placements => setSearchPlacements(placements), 300),
    [],
  );

  const onSearchChange = useCallback(e => debouncedSetSearchPlacements(e.target.value), [
    debouncedSetSearchPlacements,
  ]);

  const onFilterChange = useCallback(e => {
    setBreakdownType(e.target.value);
  }, []);

  const placements = details.placement || {};

  let breakdownData = [];
  let textFieldPlaceholder = '';
  switch (breakdownType) {
    case PLACEMENT_FILTER_TYPES.REGION:
      breakdownData = placementsByRegion;
      textFieldPlaceholder = PLACEMENT_FILTER_LABELS[breakdownType];
      break;
    case PLACEMENT_FILTER_TYPES.SENDING_IP:
      breakdownData = placementsBySendingIp;
      textFieldPlaceholder = PLACEMENT_FILTER_LABELS[breakdownType];
      break;
    default:
      breakdownData = placementsByProvider;
      textFieldPlaceholder = `${
        PLACEMENT_FILTER_LABELS[PLACEMENT_FILTER_TYPES.MAILBOX_PROVIDER]
      } or ${PLACEMENT_FILTER_LABELS[PLACEMENT_FILTER_TYPES.REGION]}`;
      break;
  }

  if (searchPlacements) {
    const searchRegex = new RegExp(_.escapeRegExp(searchPlacements), 'i');
    switch (breakdownType) {
      case PLACEMENT_FILTER_TYPES.REGION:
        breakdownData = breakdownData.filter(item => item.region.match(searchRegex));
        break;
      case PLACEMENT_FILTER_TYPES.SENDING_IP:
        breakdownData = breakdownData.filter(item => item.sending_ip.match(searchRegex));
        break;
      default:
        breakdownData = breakdownData.filter(
          item => item.region.match(searchRegex) || item.mailbox_provider.match(searchRegex),
        );
        break;
    }
  }

  return (
    <>
      <Panel>
        <Panel.Section>
          <Grid>
            <Grid.Column xs={12} sm={6} md={6}>
              <Bold>Subject Line</Bold>
              <div>{details.subject}</div>
            </Grid.Column>
            <Grid.Column xs={12} sm={6} md={6}>
              <Bold>From</Bold>
              <div>{details.from_address}</div>
            </Grid.Column>
          </Grid>
        </Panel.Section>
        <Panel.Section>
          <Grid>
            <Grid.Column xs={12} sm={4} md={4}>
              <Bold>Started</Bold>
              <div>{format(details.start_time, FORMATS.LONG_DATETIME)}</div>
            </Grid.Column>
            <Grid.Column xs={12} sm={4} md={4}>
              <Bold>Finished</Bold>
              <div>
                {details.end_time ? (
                  format(details.end_time, FORMATS.LONG_DATETIME)
                ) : (
                  <Tag className={styles.StatusTag}>{details.status}</Tag>
                )}
              </div>
            </Grid.Column>
          </Grid>
        </Panel.Section>
        <Panel.Section>
          <Bold>Test Name</Bold>
          <p>{details.test_name || 'None'}</p>
        </Panel.Section>
      </Panel>
      <div className={styles.FolderAuthenticationPanels}>
        <Panel sectioned title="Folder Placement" className={styles.FolderAuthenticationPanel}>
          <FolderPlacementBarChart placements={placements} />
        </Panel>
        <Panel sectioned title="Authentication" className={styles.FolderAuthenticationPanel}>
          <AuthenticationResults data={details.authentication} />
        </Panel>
      </div>
      <Panel title="Placement Breakdown">
        <Panel.Section>
          <div className={styles.BreakdownFilters}>
            <Select
              id="select-placement-type"
              data-id="select-placement-type"
              label="Breakdown by"
              options={PLACEMENTS_TYPE_OPTIONS}
              onChange={onFilterChange}
            />
            <TextField
              id="textfield-placement-search"
              data-id="text-field-placement-search"
              label="Filter"
              suffix={<Search />}
              onChange={onSearchChange}
              placeholder={textFieldPlaceholder}
            />
          </div>
        </Panel.Section>
        <PlacementBreakdown
          data-id="placement-breakdown-table"
          type={breakdownType}
          data={breakdownData}
        />
      </Panel>
      <OGOnlyWrapper as="div" className={styles.ClearSection} />
      <Panel sectioned title="Time to Receive Mail">
        <TimeToReceiveSection data={details.time_to_receive} />
      </Panel>
    </>
  );
};

export default TestDetails;
