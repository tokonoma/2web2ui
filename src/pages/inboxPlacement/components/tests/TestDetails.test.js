import React from 'react';
import { shallow } from 'enzyme';
import TestDetails from '../TestDetails';
import { PLACEMENT_FILTER_TYPES } from '../../constants/types';

jest.mock('lodash/debounce', () => jest.fn(fn => fn));
jest.mock('date-fns', () => ({ format: jest.fn().mockReturnValue('Jul 8th 2019 11:49am') }));
jest.mock('src/hooks/useHibanaOverride', () => styles => styles);

describe('Component: TestDetails', () => {
  const samplePlacement = {
    inbox_pct: 0,
    spam_pct: 0,
    missing_pct: 1,
  };
  const subject = ({ ...props }) => {
    const defaults = {
      details: {
        start_time: '2019-07-08T15:49:56.954Z',
        end_time: null,
        subject: 'Fooo',
        test_name: 'Baz',
      },
      placementsByProvider: [],
    };
    return shallow(<TestDetails {...defaults} {...props} />);
  };

  const searchFieldSelector = '[data-id="text-field-placement-search"]';
  const placementTypeSelectSelector = '[data-id="select-placement-type"]';
  const placementBreakdownSelector = '[data-id="placement-breakdown-table"]';

  it('renders correctly', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('renders test name as None when it does not exist', () => {
    const detailsWithoutName = {
      start_time: '2019-07-08T15:49:56.954Z',
      end_time: null,
      subject: 'Fooo',
    };
    const wrapper = subject({ details: detailsWithoutName });
    expect(wrapper).toHaveTextContent('None');
  });

  it('renders providers breakdown correctly', () => {
    const placementsByProvider = [
      {
        mailbox_provider: 'dogmail.com',
        placement: samplePlacement,
      },
    ];

    expect(
      subject({ placementsByProvider })
        .find(placementBreakdownSelector)
        .prop('data'),
    ).toEqual(placementsByProvider);
  });

  it('changes placement data when changing select', () => {
    const placementsByProvider = [
      {
        mailbox_provider: 'dogmail.com',
        placement: samplePlacement,
      },
    ];
    const placementsByRegion = [
      {
        placement: samplePlacement,
        region: 'north america',
      },
    ];
    const placementsBySendingIp = [
      {
        placement: samplePlacement,
        sending_ip: '101.101',
      },
    ];
    const wrapper = subject({ placementsByProvider, placementsByRegion, placementsBySendingIp });
    // Default data
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual(placementsByProvider);

    // Change to Region Filter
    wrapper
      .find(placementTypeSelectSelector)
      .simulate('change', { target: { value: PLACEMENT_FILTER_TYPES.REGION } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual(placementsByRegion);

    // Change to Sending Ip Filter
    wrapper
      .find(placementTypeSelectSelector)
      .simulate('change', { target: { value: PLACEMENT_FILTER_TYPES.SENDING_IP } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual(placementsBySendingIp);

    // Change back to Mailbox Provider filter
    wrapper
      .find(placementTypeSelectSelector)
      .simulate('change', { target: { value: PLACEMENT_FILTER_TYPES.MAILBOX_PROVIDER } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual(placementsByProvider);
  });

  it('changes placement data when using search bar on providers list', () => {
    const dogmail = {
      mailbox_provider: 'dogmail.net',
      region: 'north america',
      sending_ip: '101.101',
      placement: samplePlacement,
    };
    const gmail = {
      mailbox_provider: 'gmail.com',
      region: 'europe',
      sending_ip: '101.102',
      placement: samplePlacement,
    };
    const hotmail = {
      mailbox_provider: 'hotmail.com',
      region: 'global',
      sending_ip: '101.103',
      placement: samplePlacement,
    };
    const placementsByProvider = [dogmail, gmail, hotmail];

    const wrapper = subject({ placementsByProvider });

    // Match one result
    wrapper.find(searchFieldSelector).simulate('change', { target: { value: 'net' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([dogmail]);

    // Match more than one result
    wrapper.find(searchFieldSelector).simulate('change', { target: { value: 'gmail' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([dogmail, gmail]);

    // No results
    wrapper
      .find(searchFieldSelector)
      .simulate('change', { target: { value: 'this wont match anything' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([]);

    // Search by case insensitive region match
    wrapper.find(searchFieldSelector).simulate('change', { target: { value: 'america' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([dogmail]);
  });

  it('changes placement data when using search bar on regions list', () => {
    const northAmerica = {
      region: 'north america',
      placement: samplePlacement,
    };
    const southAmerica = {
      region: 'south america',
      placement: samplePlacement,
    };
    const europe = {
      region: 'europe',
      placement: samplePlacement,
    };
    const placementsByRegion = [northAmerica, southAmerica, europe];

    const wrapper = subject({ placementsByRegion });

    // Change to Region list
    wrapper
      .find(placementTypeSelectSelector)
      .simulate('change', { target: { value: PLACEMENT_FILTER_TYPES.REGION } });

    // Match one result
    wrapper.find(searchFieldSelector).simulate('change', { target: { value: 'north' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([northAmerica]);

    // Match more than one result
    wrapper.find(searchFieldSelector).simulate('change', { target: { value: 'america' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([
      northAmerica,
      southAmerica,
    ]);

    // No results
    wrapper
      .find(searchFieldSelector)
      .simulate('change', { target: { value: 'this wont match anything' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([]);
  });

  it('changes placement data when using search bar on sending ip list', () => {
    const ip1 = {
      sending_ip: '101.101',
      placement: samplePlacement,
    };
    const ip2 = {
      sending_ip: '101.102',
      placement: samplePlacement,
    };
    const ip3 = {
      sending_ip: '103.104',
      placement: samplePlacement,
    };
    const placementsBySendingIp = [ip1, ip2, ip3];

    const wrapper = subject({ placementsBySendingIp });

    // Change to Region list
    wrapper
      .find(placementTypeSelectSelector)
      .simulate('change', { target: { value: PLACEMENT_FILTER_TYPES.SENDING_IP } });

    // Match one result
    wrapper.find(searchFieldSelector).simulate('change', { target: { value: '102' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([ip2]);

    // Match more than one result
    wrapper.find(searchFieldSelector).simulate('change', { target: { value: '101' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([ip1, ip2]);

    // No results
    wrapper
      .find(searchFieldSelector)
      .simulate('change', { target: { value: 'this wont match anything' } });
    expect(wrapper.find(placementBreakdownSelector).prop('data')).toEqual([]);
  });
});
