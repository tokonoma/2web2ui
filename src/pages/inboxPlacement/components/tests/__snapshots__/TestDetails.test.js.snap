// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TestDetails renders correctly 1`] = `
<Fragment>
  <Panel>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={6}
          sm={6}
          xs={12}
        >
          <Bold>
            Subject Line
          </Bold>
          <div>
            Fooo
          </div>
        </Grid.Column>
        <Grid.Column
          md={6}
          sm={6}
          xs={12}
        >
          <Bold>
            From
          </Bold>
          <div />
        </Grid.Column>
      </Grid>
    </Panel.Section>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          sm={4}
          xs={12}
        >
          <Bold>
            Started
          </Bold>
          <div>
            Jul 8th 2019 11:49am
          </div>
        </Grid.Column>
        <Grid.Column
          md={4}
          sm={4}
          xs={12}
        >
          <Bold>
            Finished
          </Bold>
          <div>
            <Tag
              className="StatusTag"
            />
          </div>
        </Grid.Column>
      </Grid>
    </Panel.Section>
    <Panel.Section>
      <Bold>
        Test Name
      </Bold>
      <p>
        Baz
      </p>
    </Panel.Section>
  </Panel>
  <div
    className="FolderAuthenticationPanels"
  >
    <Panel
      className="FolderAuthenticationPanel"
      sectioned={true}
      title="Folder Placement"
    >
      <FolderPlacementBarChart
        placements={Object {}}
      />
    </Panel>
    <Panel
      className="FolderAuthenticationPanel"
      sectioned={true}
      title="Authentication"
    >
      <AuthenticationResults />
    </Panel>
  </div>
  <Panel
    title="Placement Breakdown"
  >
    <Panel.Section>
      <div
        className="BreakdownFilters"
      >
        <Select
          data-id="select-placement-type"
          id="select-placement-type"
          label="Breakdown by"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Mailbox Provider",
                "value": "mailbox-provider",
              },
              Object {
                "label": "Region",
                "value": "region",
              },
              Object {
                "label": "Sending IP",
                "value": "sending-ip",
              },
            ]
          }
        />
        <TextField
          data-id="text-field-placement-search"
          id="textfield-placement-search"
          label="Filter"
          onChange={[Function]}
          placeholder="Mailbox Provider or Region"
          suffix={<Search />}
        />
      </div>
    </Panel.Section>
    <PlacementBreakdown
      data={Array []}
      data-id="placement-breakdown-table"
      type="mailbox-provider"
    />
  </Panel>
  <OGOnlyWrapper
    as="div"
    className="ClearSection"
  />
  <Panel
    sectioned={true}
    title="Time to Receive Mail"
  >
    <TimeToReceiveSection />
  </Panel>
</Fragment>
`;
